clang/include/clang/AST/PreorderAST.h

```diff
@@ -0,0 +1,165 @@
+//===------- PreorderAST.h: An n-ary preorder abstract syntax tree -------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the interface for an n-ary preorder abstract syntax tree
+//  which is used to semantically compare two expressions.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_PREORDER_AST_H
+#define LLVM_CLANG_PREORDER_AST_H
+
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/CanonBounds.h"
+#include "clang/AST/Expr.h"
+
+namespace clang {
+  // Each node of the PreorderAST has a variable-count list which stores the
+  // Name of each variable along with a count of the number of times the
+  // variable appears in the node.
+  struct VarTy {
+    std::string Name;
+    llvm::APSInt Count;
+  };
+
+  // Each node of the PreorderAST contains 3 fields:
+  // - Opcode: The Opcode for the node.
+  // - variable-count list: A list of variables and their counts.
+  // - Constant value: All Constants occurring in the node are Constant folded.
+
+  // For example, consider an expression E = a + 3 + b + 4 + a
+  // A node of the PreorderAST for E would contain the following:
+  // - Opcode: +
+  // - variable-count list: [a:2, b:1]
```

Comment:
I think this AST is overly specialized to the particular transformation that you are doing.   That's fine for this change, but I think you'll want to generalize it for symbolic evaluation.
- Some operators are commutative, but not associative.  An example is signed integer arithmetic, which has undefined behavior on overflow.  You can't represent them properly in this AST.
- Some arithmetic operators are not commutative or associative, such as division.
- It would be helpful to have leaf expressions that are non-variable/non-constant expressions.